File: House_Robber_II/rob.java
class Solution {
    public int rob(int[] nums) {
        if (nums.length == 0) return 0;
        if (nums.length == 1) return nums[0];
        int rob_robf = nums[0], rob_nrobf = 0, nrob_robf = 0, nrob_nrobf = 0; // initialize from first house
        
        for (int i = 1; i < nums.length; i++) {
            int tmp = nrob_robf + nums[i];
            nrob_robf = Math.max(rob_robf, nrob_robf);
            rob_robf = tmp;
            
            tmp = nrob_nrobf + nums[i];
            nrob_nrobf = Math.max(rob_nrobf, nrob_nrobf);
            rob_nrobf = tmp;
        }
        return Math.max(rob_nrobf, nrob_robf);
    }
}

File: House_Robber_II/rob.py
class Solution(object):
    def rob(self, nums):
        if len(nums) == 0:
            return 0
        if len(nums) == 1:
            return nums[0]
        
        rob_robf = nums[0]
        rob_nrobf = 0
        nrob_robf = 0
        nrob_nrobf = 0
        
        for num in nums[1:]:
            tmp = rob_nrobf + num
            rob_nrobf = max(rob_robf, rob_nrobf)
            rob_robf = tmp
            
            tmp = nrob_nrobf + num
            nrob_nrobf = max(nrob_robf, nrob_nrobf)
            nrob_robf = tmp
        
        return max(rob_nrobf, nrob_robf)
